config/emitter.go

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package config
+
+import (
+	"sync"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// emitter enables threadsafe registration and broadcasting to configuration listeners
```

Comment:
Threadsafe, yes, but the concurrency model isn't ideal. Listeners are executed synchronously (otherwise lots of code assumptions break), so I've taken care to release all locks before issuing the callbacks (otherwise we run the rusk of deadlock with the `RWLock`.)